### PR TITLE
Add support for cracking Signal passphrases

### PR DESCRIPTION
**Pull request by kholia**
This cracks Signal Android app passphrases.

Notes, 

+  Modern versions of Signal do not support using a passphrase.

+ This was tested with Signal Android app version 4.13.5 only.

+ Run against `/data/data/org.thoughtcrime.securesms/shared_prefs/SecureSMS-Preferences.xml` file.

+ In modern version of Signal Android app (e.g. 4.19.3), the "Screen lock" code seems to be stored in plaintext in SecureSMS-Preferences.xml file

### CODE REVIEW

File: src/signal_fmt_plug.c [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+		int keylen = 16;
+		unsigned char hash[20];
+
+		pkcs12_pbe_derive_key(1, cur_salt->iterations,
+				MBEDTLS_PKCS12_DERIVE_KEY,
+				(unsigned char*)saved_key[index],
+				saved_len[index], cur_salt->mac_salt,
+				cur_salt->mac_salt_size, key, keylen);
+		hmac_sha1(key, keylen, cur_salt->master_secret, cur_salt->master_secret_size - 20, hash, 20);
+		cracked[index] = (0 == memcmp(hash, cur_salt->master_secret + cur_salt->master_secret_size - 20, 20));
```
Comment (magnumripper) [inline src/signal_fmt_plug.c]: We hates this asswards constant-first notation :-)

Either
```c
cracked[index] = (memcmp(hash, cur_salt->master_secret + cur_salt->master_secret_size - 20, 20) == 0);
```

or (this is how I'd write it, fwiw)
```c
cracked[index] = !memcmp(hash, cur_salt->master_secret + cur_salt->master_secret_size - 20, 20);
```

File: src/signal_fmt_plug.c [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+{
+	cur_salt = (struct custom_salt *)salt;
+}
+
+static int crypt_all(int *pcount, struct db_salt *salt)
+{
+	const int count = *pcount;
+	int index;
+
+	memset(cracked, 0, sizeof(cracked[0]) * cracked_count);
```
Comment (magnumripper) [inline src/signal_fmt_plug.c]: I believe this memset is pointless.

File: src/signal_fmt_plug.c [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+		BENCHMARK_LENGTH,
+		0,
+		PLAINTEXT_LENGTH,
+		BINARY_SIZE,
+		BINARY_ALIGN,
+		SALT_SIZE,
+		SALT_ALIGN,
+		MIN_KEYS_PER_CRYPT,
+		MAX_KEYS_PER_CRYPT,
+		FMT_CASE | FMT_8_BIT | FMT_OMP | FMT_HUGE_INPUT,
```
Comment (magnumripper) [inline src/signal_fmt_plug.c]: Is FMT_HUGE_INPUT really needed or is it a template leftover?

Comment (kholia) [reply src/signal_fmt_plug.c]: It was a leftover. Removed now.

File: src/signal_fmt_plug.c (hunk 2 above)
Comment (kholia) [reply src/signal_fmt_plug.c]: Yep, removed now.

File: src/signal_fmt_plug.c (hunk 1 above)
Comment (kholia) [reply src/signal_fmt_plug.c]: Ack, fixing :-)

**Pull request merged**